cmd/cstor-pool-mgmt/controller/backup-controller/new_backup_controller.go

```diff
@@ -0,0 +1,232 @@
+
+			if !IsRightCStorPoolMgmt(newbkp) {
+				return
+			}
+
+			if newbkp.ResourceVersion == oldbkp.ResourceVersion {
+				return
+			}
+
+			if IsDestroyEvent(newbkp) {
+				q.Operation = common.QOpDestroy
+				glog.Infof("BackupCstor Destroy event : %v, %v", newbkp.ObjectMeta.Name, string(newbkp.ObjectMeta.UID))
+				controller.recorder.Event(newbkp, corev1.EventTypeNormal, string(common.SuccessSynced), string(common.MessageDestroySynced))
+			} else {
+				glog.Infof("BackupCstor Modify event : %v, %v", newbkp.ObjectMeta.Name, string(newbkp.ObjectMeta.UID))
+				q.Operation = common.QOpSync
+				controller.recorder.Event(newbkp, corev1.EventTypeNormal, string(common.SuccessSynced), string(common.MessageModifySynced))
+			}
+			controller.enqueueBackupCStor(newbkp, q)
+		},
+		DeleteFunc: func(obj interface{}) {
+			bkp := obj.(*apis.BackupCStor)
+			if !IsRightCStorPoolMgmt(bkp) {
+				return
+			}
+			glog.Infof("BackupCStor Resource delete event: %v, %v", bkp.ObjectMeta.Name, string(bkp.ObjectMeta.UID))
+		},
+	})
+	return controller
+}
+
+// enqueueBackupCStor takes a BackupCStor resource and converts it into a namespace/name
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than BackupCStor.
+func (c *BackupController) enqueueBackupCStor(obj *apis.BackupCStor, q common.QueueLoad) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	q.Key = key
+	c.workqueue.AddRateLimited(q)
+}
+
+// cleanupOldBackup set fail status to old pending backup
+func cleanupOldBackup(clientset clientset.Interface) {
+	bkplabel := "cstorpool.openebs.io/uid=" + os.Getenv(string(common.OpenEBSIOCStorID))
+	bkplistop := metav1.ListOptions{
+		LabelSelector: bkplabel,
+	}
+	bkplist, err := clientset.OpenebsV1alpha1().BackupCStors(metav1.NamespaceAll).List(bkplistop)
+	if err != nil {
+		return
+	}
+
+	for _, bkp := range bkplist.Items {
+		switch bkp.Status {
+		case apis.BKPCStorStatusInProgress:
+			//Backup was in in-progress state
+			laststat := findLastBackupStat(clientset, &bkp)
+			updateBackupStatus(clientset, &bkp, laststat)
+		case apis.BKPCStorStatusDone:
+			continue
+		default:
+			//Set backup status as failed
+			updateBackupStatus(clientset, &bkp, apis.BKPCStorStatusFailed)
+		}
+	}
+}
+
+// updateBackupStatus will update the backup status to given status
+func updateBackupStatus(clientset clientset.Interface, bkp *apis.BackupCStor, status apis.BackupCStorStatus) {
+	bkp.Status = status
+
+	_, err := clientset.OpenebsV1alpha1().BackupCStors(bkp.Namespace).Update(bkp)
+	if err != nil {
+		glog.Errorf("Failed to update backup(%s) status(%s)", status, bkp.Name)
+		return
+	}
+}
+
+// findLastBackupStat will find the status of backup from last-backup
+func findLastBackupStat(clientset clientset.Interface, bkp *apis.BackupCStor) apis.BackupCStorStatus {
```

Comment:
error need to be returned for this, and caller need to handle it. Same comment applies for same fn in diff package

Reply:
ok